### ISSUE TITLE
show be more document on error handling

### ISSUE THREAD

**Issue by calidion** [balderdashy/waterline-docs#40]
and make the error handling more specific.

**Comment by dmarcelino** [comment]
Hi @calidion, can you be a bit more specific and provide examples? Thanks!

**Comment by calidion** [comment]
@dmarcelino 
Many times the model errors thrown with no detailed information on which feature is error and why it is error.

**Comment by dmarcelino** [comment]
I see, this repo is for maintaining the waterline documentation, can you please raise the issue in the [waterline core repo](https://github.com/balderdashy/waterline)? Examples would be good otherwise it will be difficult to fix the issue. Tanks.

**Comment by jurajpetrik** [comment]
I would like to see an example how to figure what kind of error was thrown, what kind of errors can be thrown and somehow transform it to a useful message for the user.
Something like( just inventing the syntax here):

    if(err.type === "validation_syntax" && err.attribute === "email"){
      res.badRequest("You must've misspelled your email");
    }
    else if (err.type === "not_found"){
      res.notFound("The user was not found. Are you registered?");
    }

I found the repo waterline-errors https://github.com/vanetix/waterline-errors but can't find any guidelines on how to use it..

**Comment by dmarcelino** [comment]
@jurajpetrik, I fear each adapter does it in its own way so it would be difficult (maybe impossible) to have such a guide (for now). https://github.com/vanetix/waterline-errors is not even maintained by @balderdashy making it even more difficult to be considered a standard...

Maybe this means we should think about creating an interface dealing with error handling that would be common among adapters.

**Comment by GochoMugo** [comment]
IMO adapters should use a common interface for defining errors. That way consumers can handle errors better. This should be prioritized highly.

**Comment by calidion** [comment]
@dmarcelino 
@GochoMugo 

I have created a way on error definition and error handling.
It is very easy to define and use, maybe you can give it a try.
https://github.com/Web-Errors/node.

errors can be defined very easily and maintainable.

**Comment by calidion** [comment]
ERROR CODE FORMAT:
 0x[31-24][23-16][15-8][7-0]
 [0-7]bit : Entity Code,
 [8-15]bit: Property / Entity Code
 [16-23]bit: Event Code
 [24-31]bit: Error Type Code

**Comment by calidion** [comment]
I use web-errors in browser side too,
it is now making the communication between the server side and the client side more smooth and universal for my projects.

The error definition can be project specific and universal.
and it has i18n support too.

**Comment by dmarcelino** [comment]
Hi @calidion, I've been reading through web-errors for nodejs README and I like the fact it comes with i18n support. The part I'm less sure is how we would replace [waterline-errors](https://github.com/vanetix/waterline-errors) for it. I'm guessing that would require some other layer?

**Comment by calidion** [comment]
@dmarcelino 

The web errrors project is developed on the intention that restful apis can be unified in error description.
So mostly it is a pure definition tools, you will get is the error names defined and the corresponding texts composed and translated. 

You may not consider it as a program:)

So from my point of view, no layer needed.

just use it directly.

```js
var errors = require('web-errors').errors;
res.json(errors.SUCCESS);
```
or
```js
var errors = require('web-errors').errors;
res.json(errors.WATERLINE_FIELD_SHOULD_BE_INTEGER);
```
where
WATERLINE_FIELD_SHOULD_BE_INTEGER
should be defined with you own error files.

you can find a full example here:
https://github.com/Web-Errors/node/tree/master/lib/data


errors are defined here:
https://github.com/Web-Errors/node/blob/master/lib/data/errors.json

and errors can be fullly customized.

**Comment by dmarcelino** [comment]
I see, it makes more sense to me now. If I'm understanding correctly we would benefit from using `web-errors` in a separate library shared by all adapters. Otherwise we would need to repeat the `data` folder in all adapters. Perhaps we can modify `waterline-errors` so it uses `web-errors`, that way we wouldn't need to change adapters already using `waterline-errors`. Does this make sense?

**Comment by devinivy** [comment]
I'd like to think of waterline as not knowing anything about HTTP, and not assume it's being used in a web API– though I think there is a reasonable mapping between waterline errors and HTTP errors for many cases.  For hapi I have a module for this called waterline-boom.  The waterline-errors repo should be the go-to for adapter errors, as I see it!  We can certainly improve that repo.

**Comment by calidion** [comment]
@dmarcelino 
Yes, web-errors is only a json object with error ids(string), error descriptions(text) and error codes(integer). and irrelavent to any thing though it is named web-errors.
which means it's main purpose is to ease the communication between the restful apis:)

**Comment by calidion** [comment]
@devinivy
I hope u don't be mislead by its name, or i should change its name when it is ready.
In fact, I am trying to make a general error project for front end, back end, database exceptions,  form validations, and restful apis which are revolving the web.
for nodejs, php, java, go, etc.
it has no direct relationship with http or tcp. it is merely a project trying to ease the way that errors be communicated between projects or companies or layers.
though it seems not that ready:)

@dmarcelino 
I hope we can work together to figure out more most frequently used general errors for other projects to start with or follow than define a new project specific error:)
which i would like to call them error prototypes.

**Comment by GochoMugo** [comment]
@calidion the idea of creating a protocol/contract in which errors can be propagated and discovered by different implementation layers, which in our case is between the ORM and the underlying database implementation layer such as mongodb, redis, disk etc., is promising.

**web-errors** should be to errors what **swagger** is to **API definitions**. While the comparison is far-fetched, having a way of discovering errors from one layer to the next would help handle them better.

It might take a long time to have such contracts in most projects to be agreed upon but we do have the opportunity to implement them in waterline and its ecosystem. As much as waterline was designed to work with sails, developing it as an independent project claiming to be an ORM, should make it be as interoperable with other projects as easy as possible.

**Comment by calidion** [comment]
@GochoMugo 
yes. I started it with a good intention. But it is beyond my controller if it will reach that point though i am trying to make it possible.
It is proved only by my project which needs a unified restful api and error reporting protocol.

General protocols for different areas are the most difficult part which I currently leave it customisable. But if we can work it out, it will ease the way how errors communicate.
And maybe we can also improve the way protocols be organized by web-errors too.
